### PR TITLE
Scoreboard: Fixed Icons sometimes missing

### PR DESCRIPTION
**Pull request by TimGoll**
This PR fixes the scoreboard being stuck open after an icon is missing. This PR does this in two levels

- adding a safety check to make sure the scoreboard never stays stuck
- fixing the root cause that causes the breakage in the first place

## The problem

Some entities register custom projectiles (like the boomerang) that are then registered as the inflictor. This breaks our current UI as I expected that the weapon is always a real TTT weapon. The change in this PR searches for an entity with the given name class, if no weapon was found.

image
image

I'm pretty sure that it fixes the two linked issues.

### CODE REVIEW

File: lua/ttt2/libraries/bodysearch.lua
Comment (saibotk) [inline lua/ttt2/libraries/bodysearch.lua:964]: 2 things here:

1. You are now not showing anything when no icon is available, maybe we should just display some kind of fallback? or at least display this one without an icon?
2. This special png handling should not be needed. Because DImage, which is used in SetIcon in SimpleIcon does already turn the string into a Material, which should work automatically with .png. See https://github.com/Facepunch/garrysmod/blob/master/garrysmod/lua/vgui/dimage.lua#L91

**Review by saibotk — CHANGES_REQUESTED** [review]

File: lua/ttt2/libraries/bodysearch.lua
Comment (TimGoll) [reply lua/ttt2/libraries/bodysearch.lua:964]: 1. But what should I show if the weapon has no icon and also no inflictor? The missing icon material?
2. But it did not work when I did pass it without an .png, so who knows

Comment (TimGoll) [reply lua/ttt2/libraries/bodysearch.lua:969]: Done the first thing, not sure what I could do about the second

Comment (saibotk) [reply lua/ttt2/libraries/bodysearch.lua:969]: Why is the searchData even providing the material instead of the filename? Seems like we may be able to change that?

Comment (TimGoll) [reply lua/ttt2/libraries/bodysearch.lua:969]: because we cache the materials and I don't want to recreate the materials based on their file

File: lua/ttt2/libraries/bodysearch.lua
Comment (ZenBre4ker) [inline lua/ttt2/libraries/bodysearch.lua:829]:
```suggestion
            local wep = util.WeaponForClass(data.wep)

            if wep then
                return wep.iconMaterial
```
umm.. Shouldnt that be like this or at least check weapon.iconMaterial?

**Review by ZenBre4ker — APPROVED** [review]

File: lua/ttt2/libraries/bodysearch.lua
Comment (TimGoll) [reply lua/ttt2/libraries/bodysearch.lua:829]: you are abolutely right, REALLY good catch. I only tested it on the boomerang of course, oof

File: lua/ttt2/libraries/bodysearch.lua
Comment (saibotk) [reply lua/ttt2/libraries/bodysearch.lua:969]: I'd rather have this fixed correctly instead of only for png images, so I'd suggest adjusting the scoreboard code.
But i would also check this out on my own and see if i can find a solution, if you don't want to.

Comment (saibotk) [reply lua/ttt2/libraries/bodysearch.lua:969]: I see, can you then just add cases for jpeg and jpg, i think these are currently the only 2 other types that are allowed in the addon whitelist.

Comment (TimGoll) [reply lua/ttt2/libraries/bodysearch.lua:969]: As a disclaimer: I don't know if any addon is actually using this information, but it could be

Comment (TimGoll) [reply lua/ttt2/libraries/bodysearch.lua:969]: I considered jpeg but couldn't think of a single case I've seen jpegs used for icons. But I can add this

Comment (TimGoll) [reply lua/ttt2/libraries/bodysearch.lua:969]: Also note, this whole function here is removed as soon as we have a new scoreboard. This only converts the new bodysearch data to the legacy table

**Comment by EntranceJew** [comment]
image
image
image
minor overcorrection

**Comment by TimGoll** [comment]
The latent commit based on a suggestion by @ZenBre4ker should solve this. Are you sure you are on that commit?

**Comment by EntranceJew** [comment]
image
what's on the branch is on the branch

**Comment by TimGoll** [comment]
I have to test this tomorrow again, but looking at the code I don't see how this could happen

**Comment by ZenBre4ker** [comment]
But its not?

**Comment by EntranceJew** [comment]
for some reason the pull i performed an hour ago to be made aware of the branch did not include a commit performed over 3 hours ago

image
image
image
image
it isn't getting stuck but the boomerang isn't displaying anymore either

**Comment by TimGoll** [comment]
Oh well, I should test stuff after merging suggestions. No idea why it is broken, but I will take a look. Thanks for testing!

**Comment by TimGoll** [comment]
okay, I fixed and tested it. It works now in both cases.

The problem here was that `util.WeaponForClass(data.wep)` already does what I want and finds the correct projectile entity, therefore my check failed. The solution therefore is to check if a cached material exists.

Maybe it would be an idea to add material caching to entities as well in the future, I'm not so sure though. Nevertheless, this is a discussion for the future, this fallback for non-cached icons is useful, even if we decide to add caching to entities as this is a good way to prevent issues for edge cases

File: lua/ttt2/libraries/bodysearch.lua
Comment (saibotk) [reply lua/ttt2/libraries/bodysearch.lua]: You've missed ".jpeg"

**Review by EntranceJew — APPROVED** [review]

**Comment by EntranceJew** [comment]
tested, couldn't find a situation that broke the scoreboard or any equipment that didn't have an attribution, LGTM

**Review by saibotk — APPROVED** [review]

**Pull request merged**
